### PR TITLE
Rust: Account for captured variables

### PR DESCRIPTION
**Pull request by hvitved**
Add AST logic for identifying captured variables.

I identified and fixed two bugs as well:
- Actually use the inner scope in `variableAccessCandInScope`, as the comment says.
- Account for `rank` being potentially non-dense, by applying a newly introduced `DenseRank` library.

### CODE REVIEW

**Review by geoffw0 — COMMENTED** [review]
Code and results LGTM.  Couple of nits.

File: shared/util/codeql/util/DenseRank.qll [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+ *   predicate getRank = rankByFirstLetter/1;
+ * }
+ *
+ * predicate denseRank = DenseRank<M>::denseRank/1;
+ * ```
+ */
+module DenseRank<DenseRankInputSig Input> {
+  private import Input
+
+  private int rankRank(Ranked r, int rnk) {
```
[the remark this answers is not in the document]
Comment (hvitved) [reply shared/util/codeql/util/DenseRank.qll:53]: It was actually supposed to be `rankRank` throughout.

[the remark this answers is not in the document]
Comment (geoffw0) [reply shared/util/codeql/util/DenseRank.qll:56]: Fine with me. :)

**Review by geoffw0 — APPROVED** [review]
LGTM.

File: shared/util/codeql/util/DenseRank.qll [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  /** A ranking context. */
+  bindingset[this]
+  class C2;
+
+  /** An element that is ranked. */
+  bindingset[this]
+  class Ranked;
+
+  /** Gets the rank of `r` in the context provided by `c1` and `c2`. */
+  int getRank(C1 c1, C2 c2, Ranked r);
```
Comment (aschackmull) [inline shared/util/codeql/util/DenseRank.qll:112]: I think perhaps I would have swapped the rank and the `Ranked` columns, so the `Ranked` would be the result, same as with the builtin aggregate.

File: shared/util/codeql/util/DenseRank.qll [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  /** An element that is ranked. */
+  bindingset[this]
+  class Ranked;
+
+  /** Gets the rank of `r` in the context provided by `c`. */
+  int getRank(C c, Ranked r);
+}
+
+/** Same as `DenseRank`, but allows for a context consisting of one element. */
+module DenseRank2<DenseRankInputSig2 Input> {
```
Comment (aschackmull) [inline shared/util/codeql/util/DenseRank.qll:80]: Would the naming perhaps be a bit better if the numbering was shifted? I.e. `DenseRank1` here to indicate a size one context?

File: shared/util/codeql/util/DenseRank.qll [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+module DenseRank3<DenseRankInputSig3 Input> {
+  private import Input
+
+  private int rankRank(C1 c1, C2 c2, Ranked r, int rnk) {
+    rnk = getRank(c1, c2, r) and
+    rnk = rank[result](int rnk0 | rnk0 = getRank(c1, c2, _) | rnk0)
+  }
+
+  /** Gets the dense rank of `r` in the context provided by `c1` and `c2`. */
+  int denseRank(C1 c1, C2 c2, Ranked r) {
```
Comment (aschackmull) [inline shared/util/codeql/util/DenseRank.qll:125]: I think perhaps I would have swapped the rank and the Ranked columns, so the Ranked would be the result, same as with the builtin aggregate.

Comment (aschackmull) [reply shared/util/codeql/util/DenseRank.qll:125]: Whether to similarly swap in the input predicate, I don't know.

**Pull request merged**